libhoth_spi.c

```diff
@@ -0,0 +1,286 @@
+{
+    if(fd < 0 || !data || !data_len)
+        return LIBHOTH_ERR_INVALID_PARAMETER;
+
+    uint8_t rd_request[5];
+    struct spi_ioc_transfer xfer[2];
+    memset(xfer, 0, sizeof xfer);
+
+    // Read OPCODE and mailbox address
+    rd_request[0] = 0x03; // Read
+    if(address_mode_4b)
+    {
+        rd_request[1] = (address >> 24) & 0xFF;
+        rd_request[2] = (address >> 16) & 0xFF;
+        rd_request[3] = (address >> 8) & 0xFF;
+        rd_request[4] = address & 0xFF;
+    
+        xfer[0].len = 5;
+    }
+    else
+    {
+        rd_request[1] = (address >> 16) & 0xFF;
+        rd_request[2] = (address >> 8) & 0xFF;
+        rd_request[3] = address & 0xFF;
+
+        xfer[0].len = 4;
+    }
+    xfer[0].tx_buf = (unsigned long)rd_request;
+    
+    // Read in data
+    xfer[1].rx_buf = (unsigned long)data;
+    xfer[1].len = data_len;
+
+    int status = ioctl(fd, SPI_IOC_MESSAGE(2), xfer);
+    if (status < 0) {
+        return LIBHOTH_ERR_FAIL;
+    }
+
+    return LIBHOTH_OK;   
+}
+
+
+int libhoth_spi_open(const struct libhoth_spi_device_init_options* options,
+                     struct libhoth_device** out) {
+  if (out == NULL || options == NULL || options->path == NULL) {
+    return LIBHOTH_ERR_INVALID_PARAMETER;
+  }
+    int status;
+    int fd;
+    struct libhoth_device* dev = NULL;
+    struct libhoth_spi_device* spi_dev = NULL;
+
+    fd = open(options->path,O_RDWR);
+    if (fd < 0) {
+        status = LIBHOTH_ERR_INTERFACE_NOT_FOUND;
+        goto err_out;
+    }
+
+    dev = calloc(1, sizeof(struct libhoth_device));
+    if (dev == NULL) {
+        status = LIBHOTH_ERR_MALLOC_FAILED;
+        goto err_out;
+    }
+
+    spi_dev = calloc(1, sizeof(struct libhoth_spi_device));
+    if (spi_dev == NULL) {
+        status = LIBHOTH_ERR_MALLOC_FAILED;
+        goto err_out;
+    }
+
+    if(options->bits)
+    {
+        const uint8_t bits = (uint8_t)options->bits;
+        if (ioctl(fd, SPI_IOC_WR_BITS_PER_WORD, bits) < 0)
+        {
+            status = LIBHOTH_ERR_FAIL;
+            goto err_out;
+        }
+    }
+
+    if(options->mode)
+    {
+        const uint8_t mode = (uint8_t)options->mode;
+        if (ioctl(fd, SPI_IOC_WR_MODE, &mode) < 0)
+        {
+            status = LIBHOTH_ERR_FAIL;
+            goto err_out;
+        }
+    }
+
+    if(options->speed)
+    {
+        const uint32_t speed = (uint32_t)options->speed;
+        if (ioctl(fd, SPI_IOC_WR_MAX_SPEED_HZ, &speed) < 0)
+        {
+            status = LIBHOTH_ERR_FAIL;
+            goto err_out;
+        }
+    }
+
+    spi_dev->fd = fd;
+    spi_dev->mailbox_address = options->mailbox;
+    spi_dev->address_mode_4b = 1;
+
+    dev->send = libhoth_spi_send_request;
+    dev->receive = libhoth_spi_receive_response;
+    dev->close = libhoth_spi_close;
+    dev->user_ctx = spi_dev;
+
+    *out = dev;
+    return LIBHOTH_OK;
+
+    err_out:
+    if (dev != NULL) {
+        free(dev);
+    }
+    if(spi_dev != NULL){
+        close(fd);
```

Comment:
This file descriptor sometimes get leaked (for example, if calloc fails). Maybe initialize it to -1 and always close it if it's anything else?

Reply:
Added initialization for file descriptor and close() if it is initialized no matter what